### DISCUSSION TITLE
template BaseModel of Model — Odd situation?

### DISCUSSION THREAD

**Discussion by EliW** [larastan/larastan#2044] [Q&A] [answered]
OK, I assume I'm doing something 'wrong' here, but, I'm working on moving our app codebase up to phpstan:6 (currently on 5) ...  I've got a file that looks like this (Minimized to only parts that should matter) ... don't ask "why" we have this, but :) ... 
```php
use Illuminate\Container\Container as Application;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Collection;

/**
 * @template BaseModel of Model
 */
abstract class BaseRepository
{
    protected Application $app;

    public function __construct(Application $app)
    {
        $this->app = $app;
    }

    /**
     * @param array<string,scalar> $search
     * @return Builder<BaseModel>
     */
    public function query(array $search = []): Builder
    {
        $query = $this->model()::query();

        if (count($search)) {
            foreach ($search as $key => $value) {
                if (in_array($key, $this->getFieldsSearchable())) {
                    $query->where($key, $value);
                }
            }
        }
        return $query;
    }

    /**
     * Retrieve all records with given filter criteria
     *
     * @param array<string,scalar> $search
     * @param ?String[] $columns
     * @return Collection<int,BaseModel>
     */
    public function all(
        array $search = [],
        ?array $columns = ['*']
    ): Collection {
        return $this->query($search)->get($columns);
    }

}
```

And in child classes I'll have something like:
```php
/**
 * @extends BaseRepository<Claim>
 */
```

The problem I'm running into, is that I'm getting this error in phpstan:
```
Method App\Repositories\BaseRepository::all()
 should return Illuminate\Database\Eloquent\Collection<int, BaseModel of Illuminate\Database\Eloquent\Model>
   but returns Illuminate\Database\Eloquent\Collection<int, Illuminate\Database\Eloquent\Model>.
```
        
If I change the return definition to `@return Collection<int,Model>` ... phpstan stops complaining.   But, that's "less correct".   It in fact WILL return a collection of BaseModel .... so in the case of `@extends BaseRepository<Claim>` version it will be a collection of `Claim` ... not of general `Model`

Not sure why phpstan/larastan are confused on this one.    If I leave it at `Model` then anything calling this won't realize that it's actually getting `Claim` instances ... so at the moment I'm leaning towards just telling phpstan to ignore this line.   

But I feel like I must have something wrong.

**Comment by szepeviktor** [comment]
`->get()` may not be generics-compatible ...

**Comment by calebdw** [comment, marked as answer]
Generally template types start with `T` (e.g., `TModel` instead of `BaseModel`)

But more to the point, this is a bug in the `EloquentBuilderExtension`---more specifically when determining what the collection class should be and populating the inner generics. This passes just fine with PHPStan: https://phpstan.org/r/295d880e-3a38-44dc-b7a2-05b9f557e087

However, for the time being it's okay to just ignore the error. What really matters is what the return type of the method is

**Reply by calebdw** [reply]
This has been fixed and published in my fork: https://github.com/calebdw/larastan/pull/16

I've added the fix to https://github.com/larastan/larastan/pull/1990

**Reply by EliW** [reply]
> Generally template types start with `T` (e.g., `TModel` instead of `BaseModel`)

*shrug* Nothing in phpstan requires that, I know that internal Laravel classes tend to use the `Txxxx` nomenclature, but it's not required at all.

> But more to the point, this is a bug in the `EloquentBuilderExtension`---more specifically when determining what the collection class should be and populating the inner generics. This passes just fine with PHPStan: https://phpstan.org/r/295d880e-3a38-44dc-b7a2-05b9f557e087

Ah ok, thanks.   So shall I file an issue/ticket here?  Or do you want the honors so that we can track it for future fixing?

**Reply by EliW** [reply]
Nevermind, I JUST saw that you added it to #1990
